File: 10_June_2nd/BSTIterator.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 
// step 1 : 
    // if there is left child, find smallest node in the left sub-tree, print the value of the node and do step 2.
    // if not, print current node and do step 2.


// step 2 : 
    // if there is right child, do step 1 with the right child node as a root of the sub-tree
    // if not, pop node from stack, print value of the node and do step 2 for the node

public class BSTIterator {
    
    private Stack<TreeNode> stack;
    private TreeNode thisNode;
    private int toDoStep;

    public BSTIterator(TreeNode root) {
        this.stack = new Stack();
        this.thisNode = root;
        this.toDoStep = 1;
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        
        if (thisNode == null || toDoStep == 2 && thisNode.right == null && stack.isEmpty()) {
            return false;
        } 
        return true;
    }

    /** @return the next smallest number */
    public int next() {
        if (this.toDoStep == 1) {
            return doStep1(thisNode);
        } else {
            return doStep2();
        }
    }
    
    private int doStep1(TreeNode root) {
        
        this.toDoStep = 2;
        if (root.left != null) {
            TreeNode temp = root;
            while(temp.left != null) {
                stack.push(temp);
                temp = temp.left;
            }
            
            thisNode = temp;
            return temp.val;
        } else {
            thisNode = root;
            return root.val;
        }
    }
    
    private int doStep2() {
        
        if (thisNode.right != null) {
            return doStep1(thisNode.right);
        } else {
            thisNode = stack.pop();
            return thisNode.val;
        }
    }
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: 10_June_2nd/Gyeongwook/BSTIterator.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.stack = []
        self.current_node = TreeNode(None)
        self.current_node.right = self.most_small_node(root)
    
    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        print self.current_node.val
        
        if self.current_node is None:
            return False
            
        return self.current_node.right is not None or len(self.stack) > 0
    
    # @return an integer, the next smallest number
    def next(self):
        if self.hasNext():
            if self.current_node.right is None:
                    self.current_node = self.stack.pop()
            else:
                self.current_node = self.most_small_node(self.current_node.right)
                
            return self.current_node.val
    		
    	return None
    	
    def most_small_node(self, root):
        current_node = root
        
        if current_node is not None:
            while current_node.left is not None:
                self.stack.append(current_node)
                current_node = current_node.left
        		
        return current_node

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())